### PR TITLE
Admins/qr

### PR DESCRIPTION
**Pull request by AzuZhou**
## What's new?
- QR Page 
​
## Closes issue(s)
- [POAP-107](https://xivis.atlassian.net/browse/POAP-107)
​
## Changes include
- [X] QR table
- [X] QR filters
- [X] Routing fix
​
## Test steps to reproduce
​
## Gif / Screenshots
​
## Self-QA Checklist
- [ ] I have tested this code
- [ ] I have updated the README (if needed)
- [ ] This branch has been rebased from dev before creating the pull-request
- [ ] **Pull-request build** has passed

### CODE REVIEW

File: client/src/api.ts [hunk 1]
```diff
@@ -327,10 +343,21 @@ export function getNotifications(
...
 
+export function getQrCodes(
+  limit?: number,
+  offset?: number,
+  status?: boolean,
+  event_id?: number
+): Promise<PaginatedQrCodes> {
+  const params = queryString.stringify({ limit, offset, status, event_id }, { sort: false });
+  return secureFetch(`${API_BASE}/qr?${params}`);
+  // return secureFetch(`http://www.mocky.io/v2/5dd6a592320000e269888c21`);
```
Comment (lndgalante) [inline client/src/api.ts]: Can we remove this commented line? Or we will continue using it ?

File: client/src/backoffice/Main.tsx [hunk 2]
```diff
@@ -16,53 +16,115 @@ import { AddressManagementPage } from './AddressManagementPage';
...
+};
+
+const WithRole: React.FC<{ roles: string[] }> = ({ roles, children }) => {
+  // TODO: Get user role (Backend WIP)
+  const userRole = 'super';
+
+  if (!roles.includes(userRole)) return null;
+
+  return <>{children}</>;
+};
```
Comment (lndgalante) [inline client/src/backoffice/Main.tsx]: Hi @niconiahi I want to change this component to a HOC so we don't have to wrap all components that needs to check for role like this:
```jsx
<WithRole roles={ROUTES.issueForEvent.roles}>
  <RoleLink route={ROUTES.issueForEvent} handleClick={closeMenu} />
</WithRole>
```

Implementation of a WithRoleHOC (word HOC on the definition isn't necessary):

```jsx
type Roles = {
  roles: string[];
};

function WithRoleHOC<Props>(WrappedComponent: React.ComponentType<Props>) {
  const hocComponent = (props: Props & Roles) => {
    const userRole = 'super';
    if (!props.roles.includes(userRole)) return null;

    return <WrappedComponent {...props} />;
  };

  return hocComponent;
}
```

Usage of WithRoleHOC:

```typescript
const LabelWithRole = WithRoleHOC(Label);
const LinkWithRole = WithRoleHOC(RoleLink);

// In render:
<LabelWithRole roles={LABELS.issueBadges.roles} label={LABELS.issueBadges}/> 

<LinkWithRole
  handleClick={closeMenu}
  route={ROUTES.issueForEvent}
  roles={ROUTES.issueForEvent.roles}
/>
```

File: client/src/backoffice/QrPage.tsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+};
+
+export interface QrCode {
+  id: number;
+  qr_hash: string;
+  claimed: boolean;
+  tx_hash: string;
+  event_id: number;
+  event: PoapEvent;
+}
```
Comment (lndgalante) [inline client/src/backoffice/QrPage.tsx]: Change `interface` for `type` just for consistency.

File: client/src/api.ts [hunk 4]
```diff
@@ -97,6 +97,22 @@ export interface PaginatedNotifications {
...
+  event_id: number;
+  event: PoapEvent;
+}
+
+export interface PaginatedQrCodes {
+  limit: number;
+  offset: number;
+  total: number;
+  codes: QrCode[];
+}
```
Comment (lndgalante) [inline client/src/api.ts]: Please change `interface` for `type` just for consistency.

File: client/src/backoffice/InboxListPage.tsx [hunk 5]
```diff
@@ -76,12 +76,17 @@ const InboxListPage: FC = () => {
...
+        PAGE_SIZE,
+        page * PAGE_SIZE,
+        notificationType,
+        event_id
+      );
+      if (!response) return;
+      setNotifications(response.notifications);
+      setTotal(response.total);
+    } catch (e) {
+      console.log(e);
```
Comment (lndgalante) [inline client/src/backoffice/InboxListPage.tsx]: We're not displaying any error messages to the user/admin would be great to have a `Toast` displaying async errors i.e `We couldn't get new notifications`

File: client/src/backoffice/Main.tsx [hunk 6]
```diff
@@ -96,20 +187,111 @@ export const BackOffice: React.FC = () => (
...
+        <Switch>
+          <Route exact path={ROUTES.admin} component={Landing} />
+
+          <Route
+            exact
+            path={ROUTES.issueForEvent.path}
+            render={() => (
+              <CheckRole roles={ROUTES.issueForEvent.roles}>
+                <IssueForEventPage />
+              </CheckRole>
```
Comment (lndgalante) [inline client/src/backoffice/Main.tsx]: We shouldn't use anymore `CheckRole` component and instead use `WithRoleHOC`.
I.e: 
```jsx
const IssueForEventPageWithRole = WithRoleHOC(IssueForEventPage)
<IssueForEventPageWithRole roles={ROUTES.issueForEvent.roles} />
```

File: client/src/backoffice/QrPage.tsx [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    let event_id = undefined;
+    if (selectedEvent !== undefined) event_id = selectedEvent > -1 ? selectedEvent : undefined;
+
+    try {
+      const response = await getQrCodes(PAGE_SIZE, page * PAGE_SIZE, status, event_id);
+      if (!response) return;
+      setQrCodes(response.codes);
+      setTotal(response.total);
+    } catch (e) {
+      console.log(e);
```
Comment (lndgalante) [inline client/src/backoffice/QrPage.tsx]: We're not displaying any error messages to the user/admin would be great to have a `Toast` displaying async errors i.e `We couldn't get new notifications`

File: client/src/backoffice/QrPage.tsx [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    } finally {
+      setIsFetchingQrCodes(false);
+    }
+  };
+
+  const handleCheckbox = (value: Value) => {
+    if (!claimStatus.includes(value)) return setClaimStatus([...claimStatus, value]);
+    setClaimStatus([...claimStatus].filter(val => val !== value));
```
Comment (lndgalante) [inline client/src/backoffice/QrPage.tsx]: We should use setState callbacks:

```typescript
if (!claimStatus.includes(value)) return setClaimStatus(prevClaimStatus => [...prevClaimStatus, value]);
setClaimStatus(prevClaimStatus => [...prevClaimStatus].filter(val => val !== value));
```

File: client/src/backoffice/QrPage.tsx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                </div>
+
+                <div className={'col-md-5 ellipsis'}>
+                  <span className={'visible-sm'}>Event: </span>
+                  {qr.event === null ? '-' : qr.event.name}
+                </div>
+
+                <div className={'col-md-2 center status'}>
+                  <span className={'visible-sm'}>Status: </span>
+                  <img src={qr.claimed ? checked : error} className={'status-icon'} alt={''} />
```
Comment (lndgalante) [inline client/src/backoffice/QrPage.tsx]: Please let's add `alternative` attribute to `img` tag.

**Review by lndgalante — CHANGES_REQUESTED** [review]
Hi @niconiahi I've left a few comments on how to improve the codebase.
Let me know if there's something not 100% clear.

Please also check and fix ESLint warnings
Captura de Pantalla 2019-11-25 a la(s) 10 46 14 a  m

File: client/src/backoffice/QrPage.tsx (hunk 7 above)
Comment (niconiahi) [reply client/src/backoffice/QrPage.tsx]: we are now displaying the error message sent from the backend team

Comment (niconiahi) [reply client/src/backoffice/QrPage.tsx]: we are now displaying the error message sent from the backend team

File: client/src/backoffice/QrPage.tsx (hunk 8 above)
Comment (niconiahi) [reply client/src/backoffice/QrPage.tsx]: should we always use callbacks when possible?

File: client/src/backoffice/QrPage.tsx (hunk 7 above)
Comment (lndgalante) [reply client/src/backoffice/QrPage.tsx]: Great now the user has at least some kind of feedback if something wrong happened.

File: client/src/backoffice/QrPage.tsx (hunk 8 above)
Comment (lndgalante) [reply client/src/backoffice/QrPage.tsx]: Yes when possible we should prefer setState hook callback and avoid using the state outside the function context.

File: client/src/backoffice/Main.tsx (hunk 2 above)
Comment (niconiahi) [reply client/src/backoffice/Main.tsx]: usage done, missing props typing

Comment (lndgalante) [reply client/src/backoffice/Main.tsx]: Great @niconiahi don't forget about https://github.com/Xivis/poap/pull/42#discussion_r350188822

**Comment by niconiahi** [comment]
completed

**Comment by lndgalante** [comment]
Thanks @niconiahi for the fixes on this PR we're going to begin merging everything into `ADMINS/main`

**Review by lndgalante — APPROVED** [review]

**Pull request merged**